translator/src/main/scala/io/kaitai/struct/testtranslator/specgenerators/JuliaSG.scala

```diff
@@ -0,0 +1,78 @@
+package io.kaitai.struct.testtranslator.specgenerators
+
+import _root_.io.kaitai.struct.ClassTypeProvider
+import _root_.io.kaitai.struct.datatype.{DataType, KSError}
+import _root_.io.kaitai.struct.exprlang.Ast
+import _root_.io.kaitai.struct.languages.JuliaCompiler
+import _root_.io.kaitai.struct.testtranslator.{Main, TestEquals, TestSpec}
+import _root_.io.kaitai.struct.translators.JuliaTranslator
+
+class JuliaSG(spec: TestSpec, provider: ClassTypeProvider) extends BaseGenerator(spec) {
+  importList.add("using Test")
+
+  val translator = new JuliaTranslator(provider, importList)
+  val className = JuliaCompiler.type2class(spec.id)
+
+  override def fileName(name: String): String = s"test_$name.jl"
+
+  override def indentStr: String = "    "
+
+  override def header(): Unit = {
+    importList.add(s"import ${className}Module")
+    out.puts
+    out.puts(s"@testset ${'"'}$className test${'"'} begin")
+    out.inc
+  }
+
+  override def runParse(): Unit = {
+    out.puts(s"r = ${className}Module.from_file(${'"'}src/${spec.data}${'"'})")
+    out.puts
+  }
+
+  override def runParseExpectError(exception: KSError): Unit = {
+    importList.add("using KaitaiStruct")
+    out.puts(s"@test_throws ${JuliaCompiler.ksErrorName(exception)} ${className}Module.from_file(${'"'}src/${spec.data}${'"'})")
+  }
+
+  override def footer(): Unit = {
+    out.dec
+    out.puts("end")
+  }
+
+  override def simpleEquality(check: TestEquals): Unit = {
+    val actStr = translateAct(check.actual)
+    val expStr = translateExp(check.expected)
+    out.puts(s"@test $actStr == $expStr")
+  }
+
+  override def floatEquality(check: TestEquals): Unit = {
+    val actStr = translateAct(check.actual)
+    val expStr = translator.translate(check.expected)
+    out.puts(s"@test $actStr ≈ $expStr atol=10^(-6)")
+  }
+
+  override def nullAssert(actual: Ast.expr): Unit = {
+    val actStr = translateAct(actual)
+    out.puts(s"@test $actStr === nothing")
+  }
+
+  override def trueArrayEquality(check: TestEquals, elType: DataType, elts: Seq[Ast.expr]): Unit =
+    simpleEquality(check)
+
+  override def testException(actual: Ast.expr, exception: KSError): Unit = {
+    importList.add("using KaitaiStruct")
+    out.puts(s"@test_throws ${JuliaCompiler.ksErrorName(exception)} ${translateAct(actual)}")
+  }
+
+  override def noAsserts() =
+    out.puts("")
+
+  def translateAct(x: Ast.expr) =
+    translator.translate(x).replace("this." + Main.INIT_OBJ_NAME, "r")
+
+  def translateExp(x: Ast.expr) =
+    translator.translate(x).replace("this._root", className)
```

Comment:
I don't like seeing this pattern at all since:

1. it uses a textual replacement... (which is inherently wrong and should be avoided as much as possible - textual replacement can be only ever appropriate in free-form text, not in anything that has a certain format that must be followed, like the code expression of some programming language. The existing `.replace("this." + Main.INIT_OBJ_NAME, "r")` is also fundamentally flawed, but at least it is very unlikely to cause any actual problems, because `val INIT_OBJ_NAME = "q1w2e3"` is sufficiently "random" that it probably won't clash with anything unintended.)
2. with something that doesn't even make sense at first sight (in general you can hardly "replace" `_root`, which is the reference to the root **instance** of the top-level class, with the **class** name itself. That is like apples and oranges).

This shouldn't be necessary, and I'm afraid it indicates some issues with the compiler code generation...